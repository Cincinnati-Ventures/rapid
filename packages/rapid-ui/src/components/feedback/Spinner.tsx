import React from 'react';
import { RapidStyles, sanitizeClassNames } from '../../utils';

type Speed = 'slow' | 'medium' | 'fast';
type Size = 'sm' | 'md' | 'lg';

interface SpinnerProps
	extends React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	> {
	styles?: string;
	label?: string;
	speed?: Speed;
	size?: Size;
}

// ------------------- TAILWINDCSS CLASSES TO FORCE JIT TO COMPILE -------------------
// spinner-slow
// spinner-medium
// spinner-fast

const Spinner = React.forwardRef<HTMLDivElement, SpinnerProps>(
	(
		{
			styles,
			speed = 'fast',
			size = 'md',
			label = 'Loading...',
			children,
			...rest
		},
		ref,
	) => {
		const defaultStyles =
			'h-6 w-6 spinner-fast rounded-full border border-solid border-black border-t-transparent border-r-transparent';
		const getSpinnerSpeed = (speed: Speed) => {
			switch (speed) {
				case 'slow':
					return 'spinner-slow';
				case 'medium':
					return 'spinner-medium';
				case 'fast':
					return 'spinner-fast';
			}
		};

		const getSpinnerSize = (size: Size) => {
			switch (size) {
				case 'sm':
					return 'h-4 w-4';
				case 'md':
					return 'h-6 w-6';
				case 'lg':
					return 'h-8 w-8';
			}
		};
		return (
			<div
				{...rest}
				ref={ref}
				className={RapidStyles(
					styles || rest.className,
					sanitizeClassNames(
						defaultStyles,
						getSpinnerSpeed(speed),
						getSpinnerSize(size),
					),
				)}
			>
				<span className='sr-only' role='alert' aria-busy='true'>
					{label}
				</span>
			</div>
		);
	},
);

export default Spinner;
