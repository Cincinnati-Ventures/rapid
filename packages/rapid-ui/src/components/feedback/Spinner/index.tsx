export { default as Spinner, spinnerSafeList } from './Spinner';
