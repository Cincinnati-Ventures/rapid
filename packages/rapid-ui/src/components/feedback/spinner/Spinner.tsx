import React, { useMemo } from 'react';
import { RapidStyles, sanitizeClassNames } from '../../../utils';

type Speed = 'slow' | 'medium' | 'fast';
type Size = 'sm' | 'md' | 'lg';

interface SpinnerProps
	extends React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	> {
	styles?: string;
	label?: string;
	speed?: Speed;
	size?: Size;
}

// ------------------- TAILWINDCSS CLASSES TO FORCE JIT TO COMPILE -------------------
export const spinnerSafeList = [
	'spinner-slow',
	'spinner-medium',
	'spinner-fast',
];
const defaultStyles =
	'h-6 w-6 spinner-fast rounded-full border border-solid border-black border-t-transparent border-r-transparent';

const Spinner = React.forwardRef<HTMLDivElement, SpinnerProps>(
	(
		{
			styles,
			speed = 'fast',
			size = 'md',
			label = 'Loading...',
			children,
			...rest
		},
		ref,
	) => {
		const getSpinnerSpeed = useMemo(() => {
			switch (speed) {
				case 'slow':
					return 'spinner-slow';
				case 'medium':
					return 'spinner-medium';
				case 'fast':
					return 'spinner-fast';
			}
		}, [speed]);

		const getSpinnerSize = useMemo(() => {
			switch (size) {
				case 'sm':
					return 'h-4 w-4';
				case 'md':
					return 'h-6 w-6';
				case 'lg':
					return 'h-8 w-8';
			}
		}, [size]);

		return (
			<div
				{...rest}
				ref={ref}
				className={RapidStyles(
					styles || rest.className,
					sanitizeClassNames(
						defaultStyles,
						getSpinnerSpeed,
						getSpinnerSize,
					),
				)}
			>
				<span className='sr-only' role='alert' aria-busy='true'>
					{label}
				</span>
			</div>
		);
	},
);

export default Spinner;
