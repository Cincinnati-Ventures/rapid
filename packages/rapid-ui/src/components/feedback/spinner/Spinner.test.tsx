import React from 'react';
import { render } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import Spinner from './Spinner';

describe('Spinner', () => {
	it('should render default correctly', () => {
		const { container } = render(<Spinner />);
		expect(container).toMatchSnapshot();
	});

	it('should render correctly with non default props', () => {
		const { container } = render(
			<Spinner
				styles='custom-class'
				size='sm'
				speed='slow'
				label='Some label'
			/>,
		);
		expect(container).toMatchSnapshot();
	});

	it('should render correctly with a couple props', () => {
		const { container } = render(<Spinner size='lg' speed='fast' />);
		expect(container).toMatchSnapshot();
	});
});
