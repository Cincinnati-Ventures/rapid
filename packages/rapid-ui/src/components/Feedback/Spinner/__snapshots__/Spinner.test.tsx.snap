// Vitest Snapshot v1

exports[`Spinner > should render correctly with a couple props 1`] = `
<div>
  <div
    class="spinner-fast rounded-full border border-solid border-black border-t-transparent border-r-transparent spinner-fast h-8 w-8"
  >
    <span
      aria-busy="true"
      class="sr-only"
      role="alert"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`Spinner > should render correctly with non default props 1`] = `
<div>
  <div
    class="spinner-fast rounded-full border border-solid border-black border-t-transparent border-r-transparent spinner-slow h-4 w-4 custom-class"
  >
    <span
      aria-busy="true"
      class="sr-only"
      role="alert"
    >
      Some label
    </span>
  </div>
</div>
`;

exports[`Spinner > should render default correctly 1`] = `
<div>
  <div
    class="spinner-fast rounded-full border border-solid border-black border-t-transparent border-r-transparent spinner-fast h-6 w-6"
  >
    <span
      aria-busy="true"
      class="sr-only"
      role="alert"
    >
      Loading...
    </span>
  </div>
</div>
`;
